src/js/signin/containers/Main.jsx

```diff
@@ -0,0 +1,167 @@
+    this.getUserData = getUserData;
+  }
+
+  componentDidMount() {
+    if (sessionStorage.userToken) {
+      this.getLogoutUrl();
+    }
+    this.getLoginUrl();
+    this.getVerifyUrl();
+    addEvent(window, 'message', (evt) => {
+      this.setMyToken(evt);
+    });
+    window.onload = this.checkTokenStatus();
+  }
+
+  componentWillUnmount() {
+    this.loginUrlRequest.abort();
+    this.verifyUrlRequest.abort();
+    this.logoutUrlRequest.abort();
+  }
+
+  getLoginUrl() {
+    this.loginUrlRequest = getLoginUrl(this.props.onUpdateLoginUrl);
+  }
+
+  getVerifyUrl() {
+    this.verifyUrlRequest = fetch(`${environment.API_URL}/v0/sessions/new?level=3`, {
+      method: 'GET',
+    }).then(response => {
+      return response.json();
+    }).then(json => {
+      this.props.onUpdateVerifyUrl(json.authenticate_via_get);
+    });
+  }
+
+  setMyToken(event) {
+    if (event.data === sessionStorage.userToken) {
+      this.getUserData(this.props.dispatch);
+      this.getLogoutUrl();
+    }
+  }
+
+  getLogoutUrl() {
+    this.logoutUrlRequest = fetch(`${environment.API_URL}/v0/sessions`, {
+      method: 'DELETE',
+      headers: new Headers({
+        Authorization: `Token token=${sessionStorage.userToken}`
+      })
+    }).then(response => {
+      return response.json();
+    }).then(json => {
+      this.props.onUpdateLogoutUrl(json.logout_via_get);
+    });
+  }
+
+  handleLogin() {
+    this.loginUrlRequest = handleLogin(this.props.login.loginUrl, this.props.onUpdateLoginUrl);
+  }
+
+  handleSignup() {
+    window.dataLayer.push({ event: 'register-link-clicked' });
+    const myLoginUrl = this.props.login.loginUrl;
+    if (myLoginUrl) {
+      window.dataLayer.push({ event: 'register-link-opened' });
+      const receiver = window.open(`${myLoginUrl}&op=signup`, '_blank', 'resizable=yes,scrollbars=1,top=50,left=500,width=500,height=750');
+      receiver.focus();
+    }
+  }
+
+  handleLogout() {
+    window.dataLayer.push({ event: 'logout-link-clicked' });
+    const myLogoutUrl = this.props.login.logoutUrl;
+    if (myLogoutUrl) {
+      window.dataLayer.push({ event: 'logout-link-opened' });
+      const receiver = window.open(myLogoutUrl, '_blank', 'resizable=yes,scrollbars=1,top=50,left=500,width=500,height=750');
+      receiver.focus();
+    }
+  }
+
+  checkTokenStatus() {
+    if (sessionStorage.userToken) {
+      if (moment() > moment(sessionStorage.entryTime).add(45, 'm')) {
+        // TODO(crew): make more customized prompt.
+        if (confirm('For security, you’ll be automatically signed out in 2 minutes. To stay signed in, click OK.')) {
+          this.handleLogin();
+        } else {
+          this.handleLogout();
+        }
+      } else {
+        if (this.getUserData(this.props.dispatch)) {
+          this.props.onUpdateLoggedInStatus(true);
+        }
+      }
+    } else {
+      this.props.onUpdateLoggedInStatus(false);
+    }
+  }
+
+  render() {
+
+    if (this.props.verify) {
```

Comment:
Where is `this.props.verify` coming from?

Reply:
Instead of returning the following in a separate conditional block, could just return one block:

```
const currentlyLoggedIn =
  this.props.verify
  ? undefined
  : this.props.login.currentlyLoggedIn;

return (
  <SignIn
    currentlyLoggedIn={currentlyLoggedIn}
    handleSignup={...}
    handleLogin={...}/>
);
```

Reply:
A little explanation here—things might have gotten a little mixed up with all the branching. I am making the `Verify` and `Signin` pages into components, while the Main container holding the shared logic around fetching the tokens, urls, etc.  So the `verify` prop on `Main` tells it whether to render the Verify page.

Would it be helpful to pull in the `Verify` page branch into this PR and review it together? There are a bunch of other changes and logic improvements on the Verify branch